### ISSUE TITLE
Iteration exhaustion branch of asynchronous for loop is shown as missing

### ISSUE THREAD

**Issue by manueljacob** [coveragepy/coveragepy#1158]
```python
import asyncio

async def async_gen():
    yield 1

async def async_test():
    a = None
    async for i in async_gen():
        print(i)

asyncio.run(async_test())
```

I’ve run `coverage run --branch script.py` with Python 3.9.2 + coverage 5.5 and from Git commit 40c87e08a98dd06dde8781bad32876b01ce9ea3b.

I would expect that no line or branch is shown as missing, however branch `8->exit` is shown as missing.

Similarly, when using the following script, branch `8->11` is shown as missing, although running the script fails at line 11.

```python
import asyncio

async def async_gen():
    yield 1

async def async_test():
    a = None
    async for i in async_gen():
        print(i)
    else:
        zzz

asyncio.run(async_test())
```

When removing the `a = None` line or running on PyPy3.7 v7.3.4, no line or branch is shown as missing.

Also, with the following synchronous version, no line or branch is shown as missing:

```python
def sync_gen():
    yield 1

def sync_test():
    a = None
    for i in sync_gen():
        print(i)

sync_test()
```

**Comment by jorants** [comment]
Came across the same problem when using coverage on [beanie](https://github.com/roman-right/beanie) which uses a lot of `async for` loops. Might be related to  #1176

**Comment by nedbat** [comment]
I added [a test for this on a branch](https://github.com/nedbat/coveragepy/tree/nedbat/bug1158), and it seems to do the right things for these Python versions: 3.7.10, 3.8.10, and 3.10.0b3.
It doesn't trace correctly on Python 3.9.5.
So this seems like a problem in Python 3.9.  Can you confirm?

**Comment by jorants** [comment]
I tried the same tests from beanie on 3.8 and there it does seem to work correctly, i can not easily test the others right now.

**Comment by nedbat** [comment]
I've reported this against CPython https://bugs.python.org/issue44621.

**Comment by nedbat** [comment]
This CPython bug has been fixed in 3.9.7.
